jxl/src/features/patches.rs

```diff
@@ -0,0 +1,290 @@
+// blend_channel value matches itself.
+// If using kBlendAbove, new is the patch and old is the original image; if
+// using kBlendBelow, the meaning is inverted.
+const PATCH_BLEND_MODE_BLEND_ABOVE: u8 = 4;
+const PATCH_BLEND_MODE_BLEND_BELOW: u8 = 5;
+// The new values (in the crop) are added to the old ones if alpha>0:
+// For first alpha channel: sample = sample = old + new * (1 - old)
+// For other channels: sample = old + alpha * new
+const PATCH_BLEND_MODE_ALPHA_WEIGHTED_ADD_ABOVE: u8 = 6;
+const PATCH_BLEND_MODE_ALPHA_WEIGHTED_ADD_BELOW: u8 = 7;
+const PATCH_BLEND_MODE_NUM_BLEND_MODES: u8 = 8;
+
+#[inline]
+fn uses_alpha(blend_mode: u8) -> bool {
+    blend_mode == PATCH_BLEND_MODE_BLEND_ABOVE
+        || blend_mode == PATCH_BLEND_MODE_BLEND_BELOW
+        || blend_mode == PATCH_BLEND_MODE_ALPHA_WEIGHTED_ADD_ABOVE
+        || blend_mode == PATCH_BLEND_MODE_ALPHA_WEIGHTED_ADD_BELOW
+}
+
+#[inline]
+fn uses_clamp(blend_mode: u8) -> bool {
+    uses_alpha(blend_mode) || blend_mode == PATCH_BLEND_MODE_MUL
+}
+
+#[derive(Debug, Clone, Copy)]
+struct PatchBlending {
+    mode: u8,
+    alpha_channel: usize,
+    clamp: bool,
+}
+
+#[derive(Debug, Clone, Copy)]
+pub struct PatchReferencePosition {
+    // Not using `ref` like in the spec here, because it is a keyword.
+    reference: usize,
+    x0: usize,
+    y0: usize,
+    xsize: usize,
+    ysize: usize,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub struct PatchPosition {
+    x: usize,
+    y: usize,
+    ref_pos_idx: usize,
+}
+
+#[derive(Debug, Default)]
+pub struct PatchesDictionary {
+    pub positions: Vec<PatchPosition>,
+    ref_positions: Vec<PatchReferencePosition>,
+    blendings: Vec<PatchBlending>,
+    blendings_stride: usize,
+}
+
+impl PatchesDictionary {
+    #[instrument(level = "debug", skip(br), ret, err)]
+    pub fn read(
+        br: &mut BitReader,
+        xsize: usize,
+        ysize: usize,
+        decoder_state: &DecoderState,
+    ) -> Result<PatchesDictionary> {
+        let num_extra_channels = decoder_state.extra_channel_info().len();
+        let blendings_stride = num_extra_channels + 1;
+        let patches_histograms = Histograms::decode(NUM_PATCH_DICTIONARY_CONTEXTS, br, true)?;
+        let mut patches_reader = patches_histograms.make_reader(br)?;
+        let num_ref_patch = patches_reader.read(br, NUM_REF_PATCH_CONTEXT)? as usize;
+        let num_pixels = xsize * ysize;
+        let max_ref_patches = 1024 + num_pixels / 4;
+        let max_patches = max_ref_patches * 4;
+        let _max_blending_infos = max_patches * 4;
+        if num_ref_patch > max_ref_patches {
+            return Err(Error::PatchesTooMany(num_ref_patch, max_ref_patches));
+        }
+        let mut total_patches = 0;
+        let mut next_size = 1;
+        let mut positions: Vec<PatchPosition> = Vec::new();
+        let mut blendings = Vec::new();
+        let mut ref_positions: Vec<PatchReferencePosition> = Vec::with_capacity(num_ref_patch);
```

Comment:
new() + try_reserve?

(we could also impl try_with_capacity() in util, I guess we use the pattern more than enough)

Reply:
done. 
Will make a pr with try_with_capacity (and remove a few places where we currently only do `with_capacity` next).
Leaving this unresolved, so I don't forget